src/parse.jl

```diff
@@ -68,20 +68,18 @@ const RULES = TupleDict(
 )
 
 # same as above, but valid for partially completed string commands. This is to determine when the key stack should be cleared.
-const PARTIAL_RULES = TupleDict(
-    # insert commands
-    r"^(?<c>[aAiIoO])" => InsertCommand,
-    # Special case: `0` is a motion command:
-    "^0\$" |> Regex => ZeroCommand,
-    # synonym commands
-    "^(?<n1>$REPEAT)(?<c>[xXCS])" |> Regex => SynonymCommand,
-    "^(?<n1>$REPEAT)($MOTION)" |> Regex => SimpleMotionCommand,
-    "^(?<n1>$REPEAT)((?|$(complex_motion())))" |> Regex => CompositeMotionCommand,
-    "^(?<n1>$REPEAT)(?<op>$OPERATOR)(?<n2>$REPEAT)(?|($TEXTOBJECT)|($MOTION))" |> Regex => OperatorCommand,
-    "^(?<n1>$REPEAT)(?<op>$OPERATOR)(?<n2>$REPEAT)((?|$(complex_motion())))" |> Regex => OperatorCommand,
-    "^(?<n1>$REPEAT)(?<op>$OPERATOR)(\\k<op>)" |> Regex => LineOperatorCommand,
-    "^(?<n1>$REPEAT)r(.)" |> Regex => ReplaceCommand,
+const PARTIAL_RULES = (
+    "^(?<c>$INSERTCHARS)?\$" |> Regex,  # InsertCommand
+    "^0?\$" |> Regex,  # ZeroCommand
+    "^(?<n1>$REPEAT)(?<c>$DELETECHARS)?\$" |> Regex,  # SynonymCommand
+    "^(?<n1>$REPEAT)($MOTION)?\$" |> Regex,  # SimpleMotionCommand
+    "^(?<n1>$REPEAT)((?|$(complex_motion(true))))?\$" |> Regex,  # CompositeMotionCommand
+    "^(?<n1>$REPEAT)((?<op>$OPERATOR)((?<n2>$REPEAT)((?|($PARTIALTEXTOBJECT)|($MOTION)))?)?)?\$" |> Regex,  # OperatorCommand
```

Comment:
Okay, so after looking over these for a while I think this is correct. Just writing out my thoughts as I get them here. 

The outermost `$REPEAT` already has a nested `?` within it, so doesn't require a new one, which you've done. 

I see you've appended a terminal `$` to each line, and on first look I didn't realize why, but on second thought it's clearly needed to invalidate a string which may have a valid command as a substring.

Reply:
Right. Perhaps I should wrap the `$REPEAT` with a `?` for consistency? In this case it would be one `()?` around the entire regexp, in between `^` and `$`.

Reply:
No, I think leaving `$REPEAT` without a `?` is alright. A little less consistent but more correct in my opinion